response.go

```diff
@@ -103,185 +156,375 @@ func (info *SQLInfo) DecodeMsgpack(d *msgpack.Decoder) error {
 	return nil
 }
 
-func (resp *Response) smallInt(d *msgpack.Decoder) (i int, err error) {
-	b, err := resp.buf.ReadByte()
+func smallInt(d *msgpack.Decoder, buf *smallBuf) (i int, err error) {
+	b, err := buf.ReadByte()
 	if err != nil {
 		return
 	}
 	if b <= 127 {
 		return int(b), nil
 	}
-	resp.buf.UnreadByte()
+	buf.UnreadByte()
 	return d.DecodeInt()
 }
 
-func (resp *Response) decodeHeader(d *msgpack.Decoder) (err error) {
+func decodeHeader(d *msgpack.Decoder, buf *smallBuf) (Header, iproto.Type, error) {
 	var l int
-	d.Reset(&resp.buf)
+	var code int
+	var err error
+	d.Reset(buf)
 	if l, err = d.DecodeMapLen(); err != nil {
-		return
+		return Header{}, 0, err
 	}
+	decodedHeader := Header{}
```

Comment:
It is meaningful constant, let it be obvious.
```suggestion
	decodedHeader := Header{Error: ErrorNo}
```